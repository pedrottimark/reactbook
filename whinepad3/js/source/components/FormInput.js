/* @flow */

import React, {Component} from 'react';

import Rating from './Rating';
import Suggest from './Suggest';

import type { FieldType, FieldValue } from '../reducers/fields';

export type Props = {
  type: FieldType,
  defaultValue?: FieldValue,
  id?: string,
  options?: Array<string>,
  label?: string,
};

class FormInput extends Component {

  props: Props;

  static defaultProps = {
    type: 'input',
  };

  getValue(): FieldValue {
    return 'value' in this.refs.input
      ? this.refs.input.value
      : this.refs.input.getValue();
  }

  render() {
    const { defaultValue, id, type } = this.props;
    const common: Object = {
      id,
      ref: 'input',
      defaultValue,
    };
    switch (type) {
      case 'year':
        return (
          <input
            {...common}
            type="number"
            defaultValue={defaultValue || new Date().getFullYear()} />
        );
      case 'suggest':
        return <Suggest {...common} options={this.props.options} />;
      case 'rating':
        return (
          <Rating {...common} />
        );
      case 'text':
        return <textarea {...common} />;
      default:
        return <input {...common} type="text" />;
    }
  }
}

export default FormInput
