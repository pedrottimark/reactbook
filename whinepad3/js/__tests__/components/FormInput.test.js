import React from 'react';
import TestUtils from 'react-addons-test-utils';
import renderer from 'react-test-renderer';

import FormInput from '../../source/components/FormInput';

describe('FormInput with no type prop', () => {

  it('renders <input type="text" />', () => {
    const tree = renderer.create(
      <FormInput />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders <input type="text" /> with defaultValue', () => {
    const tree = renderer.create(
      <FormInput defaultValue="default value" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('returns defaultValue or changed value', () => {
    const defaultValue = 'default value';
    const otherValue = 'other value';
    const component = TestUtils.renderIntoDocument(
      <FormInput defaultValue={defaultValue} />
    );
    expect(component.getValue()).toEqual(defaultValue);
    const input = TestUtils.findRenderedDOMComponentWithTag(component, 'input');
    input.value = otherValue;
    expect(component.getValue()).toEqual(otherValue);
  });

});

describe('FormInput with type="year" prop', () => {

  it('renders <input type="number" />', () => {
    const tree = renderer.create(
      <FormInput type="year" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders <input type="number" /> with defaultValue', () => {
    const tree = renderer.create(
      <FormInput type="year" defaultValue={2016} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('returns defaultValue or changed value', () => {
    const defaultValue = 2016;
    const otherValue = 2015;
    const component = TestUtils.renderIntoDocument(
      <FormInput type="year" defaultValue={defaultValue} />
    );
    expect(component.getValue()).toEqual(defaultValue);
    const input = TestUtils.findRenderedDOMComponentWithTag(component, 'input');
    input.value = otherValue;
    expect(component.getValue()).toEqual(otherValue);
  });

});

describe('FormInput with type="suggest" prop', () => {

  it('renders <div> containing <input /> and <datalist>', () => {
    const tree = renderer.create(
      <FormInput type="suggest" id="color" options={['r', 'g', 'b']} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders <div> containing <input /> and <datalist> with defaultValue', () => {
    const tree = renderer.create(
      <FormInput type="suggest" id="color" options={['r', 'g', 'b']} defaultValue="g" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('FormInput with type="rating" prop', () => {

  it('renders <div> containing <span> for each star and <input type="hidden" />', () => {
    const tree = renderer.create(
      <FormInput type="rating" id="rating" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('returns defaultValue prop of FormInput with type="rating"', () => {
    const tree = renderer.create(
      <FormInput type="rating" id="rating" defaultValue={3} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('FormInput with type="text" prop', () => {

  it('renders <textarea />', () => {
    const tree = renderer.create(
      <FormInput type="text" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders <textarea /> with defaultValue', () => {
    const defaultValue = `default
value`;
    const tree = renderer.create(
      <FormInput type="text" defaultValue={defaultValue} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('returns defaultValue or changed value', () => {
    const defaultValue = `default
value`;
    const otherValue = `other
value`;
    const component = TestUtils.renderIntoDocument(
      <FormInput type="text" defaultValue={defaultValue} />
    );
    expect(component.getValue()).toEqual(defaultValue);
    const textarea = TestUtils.findRenderedDOMComponentWithTag(component, 'textarea');
    textarea.value = otherValue;
    expect(component.getValue()).toEqual(otherValue);
  });

});
