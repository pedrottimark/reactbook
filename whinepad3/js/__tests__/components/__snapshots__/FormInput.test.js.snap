exports[`FormInput with no type prop renders <input type="text" /> 1`] = `
<input
  defaultValue={undefined}
  type="text" />
`;

exports[`FormInput with no type prop renders <input type="text" /> with defaultValue 1`] = `
<input
  defaultValue="default value"
  type="text" />
`;

exports[`FormInput with type="rating" prop renders <div> containing <span> for each star and <input type="hidden" /> 1`] = `
<div
  className="Rating"
  onMouseOut={[Function]}>
  <span
    className={null}
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <span
    className={null}
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <span
    className={null}
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <span
    className={null}
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <span
    className={null}
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <input
    type="hidden"
    value={0} />
</div>
`;

exports[`FormInput with type="rating" prop returns defaultValue prop of FormInput with type="rating" 1`] = `
<div
  className="Rating"
  onMouseOut={[Function]}>
  <span
    className="RatingOn"
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <span
    className="RatingOn"
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <span
    className="RatingOn"
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <span
    className={null}
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <span
    className={null}
    onClick={[Function]}
    onMouseOver={[Function]}>
    ☆
  </span>
  <input
    type="hidden"
    value={3} />
</div>
`;

exports[`FormInput with type="suggest" prop renders <div> containing <input /> and <datalist> 1`] = `
<div>
  <input
    defaultValue={undefined}
    list="color"
    onChange={[Function]} />
  <datalist
    id="color">
    <option
      value="r" />
    <option
      value="g" />
    <option
      value="b" />
  </datalist>
</div>
`;

exports[`FormInput with type="suggest" prop renders <div> containing <input /> and <datalist> with defaultValue 1`] = `
<div>
  <input
    defaultValue="g"
    list="color"
    onChange={[Function]} />
  <datalist
    id="color">
    <option
      value="r" />
    <option
      value="g" />
    <option
      value="b" />
  </datalist>
</div>
`;

exports[`FormInput with type="text" prop renders <textarea /> 1`] = `
<textarea
  defaultValue={undefined} />
`;

exports[`FormInput with type="text" prop renders <textarea /> with defaultValue 1`] = `
<textarea
  defaultValue="default
value" />
`;

exports[`FormInput with type="year" prop renders <input type="number" /> 1`] = `
<input
  defaultValue={2016}
  type="number" />
`;

exports[`FormInput with type="year" prop renders <input type="number" /> with defaultValue 1`] = `
<input
  defaultValue={2016}
  type="number" />
`;
